### PR TITLE
Improve the handling of quantized weights

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

Handling of quantized weights was split between two mechanisms:

- For quantized checkpoints, we used the new weight loader infrastructure.
- For quantization while loading (EETQ, FP8, bitsandbytes) we instead relied on conditional in `get_linear`.

Weight loaders support context managers to selectively load particular layers with different weight loaders, which is useful for models like Idefics2 AWQ, which uses a quantized text model, but unquantized vision and connector models. However, the context manager would be overrided by `get_linear`, which string-checks `quantizer`. Also, the context manager would not work with EETQ, FP8, and bitsandbytes.

This change migrates all quantizers to the weight loader infrastructure. This has several benefits:

- We can use context managers with all quantizers.
- All the implementation details move down to the quantizer layers, `get_linear` does not need to know how to handle quantizer linear layers.
- All quantizer weights are strongly typed, we don't pass around raw tensors.
- We don't have to pass around the `quantizer` string everywhere.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: integration-tests/models/__snapshots__/test_chat_llama/test_flash_llama_simple.json [hunk 1]
```diff
@@ -5,19 +5,19 @@
       "index": 0,
       "logprobs": null,
       "message": {
-        "content": "As of your last question, the weather in Brooklyn, New York, is typically hot and humid throughout the year. The suburbs around New York City are jealously sheltered, and at least in the Lower Bronx, there are very few outdoor environments to explore in the middle of urban confines. In fact, typical times for humidity levels in Brooklyn include:\n\n- Early morning: 80-85% humidity, with occas",
+        "content": "As of your last question, the weather in Brooklyn, New York, is typically moderate to warm year-round. The suburban areas around the borough are jealously sheltered from the Northeastern United States' harsh wind and rain systems. In fact, Brooklyn vs the urban confines of Manhattan or Staten Island is energized by an idyllic summer that often sees crisp East Coast air embraced by the mild atmosphere, year after",
```
Comment (OlivierDehaene) [inline integration-tests/models/__snapshots__/test_chat_llama/test_flash_llama_simple.json:8]: Do you know why this changed?

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py [hunk 2]
```diff
@@ -336,9 +338,21 @@ def __init__(self, index, prefix, config, weights):
...
-            prefix=f"{prefix}.mlp", config=config, weights=weights, index=index
-        )
+
+        # Only use FP8 when we are not using another quantizer. We should probably
+        # have nicer way of querying this, but for now it'll do.
+        weights_loader = weights.weights_loader
+        if (
+            can_use_fp8()
+            and isinstance(weights_loader, DefaultWeightsLoader)
+            and weights_loader.weight_class is UnquantizedWeight
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/custom_modeling/flash_llama_modeling.py:348]: Doesn't this mean that we will always use fp8 for the MLP? I don't think we want that.
In my opinion it should be the other way around; wrap all the other linear layers with a context manager using UnquantizedWeight if we are in FP8 but that's a bit cumbersome.

I guess that's why the outputs changed.

**Review by OlivierDehaene — DISMISSED** [review]

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py [hunk 3]
```diff
@@ -470,12 +492,15 @@ class FlashLlamaForCausalLM(torch.nn.Module):
     def __init__(self, prefix: str, config, weights):
         super().__init__()
 
-        self.embed_tokens = TensorParallelEmbedding(
-            prefix=(
-                "model.embed_tokens" if not prefix else f"{prefix}.model.embed_tokens"
-            ),
-            weights=weights,
-        )
+        with no_fp8(weights):
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/custom_modeling/flash_llama_modeling.py:495]: I think you are also missing the `lm_head`? or is not necessary?

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py (hunk 2 above)
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py]: Changed around as you suggested (do nothing, unless `--quantize=fp8`, then exclude attention and piece embeddings).

File: server/text_generation_server/layers/fp8.py [hunk 4]
```diff
@@ -1,7 +1,17 @@
-from enum import Enum, auto
+from dataclasses import dataclass
 
 import torch
 from text_generation_server.utils.import_utils import SYSTEM
+from text_generation_server.utils.weights import Weight
+
+
+def can_use_fp8():
```
Comment (OlivierDehaene) [inline server/text_generation_server/layers/fp8.py:8]: I don't think this is used anymore

File: server/text_generation_server/models/custom_modeling/flash_llama_modeling.py (hunk 3 above)
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_llama_modeling.py:495]: Added.

File: server/text_generation_server/layers/fp8.py (hunk 4 above)
Comment (danieldk) [reply server/text_generation_server/layers/fp8.py]: Indeed, removed.

**Review by OlivierDehaene — APPROVED** [review]
Nice!

**Pull request merged**